### DISCUSSION TITLE
Finalise Signature

### DISCUSSION THREAD

**Discussion by sshmatrix** [namesys-eth/ccip2-eth-resources#17] [General]
> Finalise Signature formatting

INPS Keygen: 

```js
let statement = `Requesting signature for IPNS key generation\n\nUsername: ${ens}\nSigned By: ${caip10}`
```

Record Signatures:

```js
string memory _req = string.concat(
    "Requesting signature for off-chain ENS record\n",
    //"\nIMPORTANT: Please verify the integrity and authenticity of connected Off-chain ENS Records Manager before signing this message\n",
    "\nENS Domain: ",
    _domain,
    "\nRecord Type: ",
    _jsonPath,
    "\nRecord Hash: 0x",
    bytesToString(abi.encodePacked(keccak256(result)), 0),
    "\nSigned By: eip155:1:",
    toChecksumAddress(_signer)
);
```

**Comment by 0xc0de4c0ffee** [comment]
```
bytes memory _result = abi.encode(_addr);
bytes memory _response = abi.encodePacked(resolver.__callback.selector, abi.encode(_signer, _signature, _result));
```
signature request format 
```
string memory _req = string.concat(
    "Requesting signature for off-chain ENS record\n",
    //"\nIMPORTANT: Please verify the integrity and authenticity of connected Off-chain ENS Records Manager before signing this message\n",
    "\nENS Domain: ",
    _domain,
    "\nRecord Type: ",
    _jsonPath,
    "\nRecord Hash: 0x",
    bytesToString(abi.encodePacked(keccak256(result)), 0),
    "\nSigned By: eip155:1:",
    toChecksumAddress(_signer)
);
```

**Comment by sshmatrix** [comment]
@0xc0de4c0ffee 

Can you also confirm this format for keygen for the final version:

```js
let statement = `Requesting signature for IPNS key generation\n\nUsername: ${ens}\nSigned By: ${caip10}`
```

So far I have put username as `domain.eth` and the user can choose the password/secret-key-identifier. I found it easier than to remember two things: username + password. What do you think?

**Reply by 0xc0de4c0ffee** [reply]
🤔 

let statement = `Requesting Signature For Off-Chain ENS Records Manager\n\nENS Domain: ${ens}\nExtradata: 0x${<keccak256(ens, password?password:"", caip10)>}\nSigned By: ${caip10}` * check https://github.com/orgs/namesys-eth/discussions/18 too 

2x hkdf to generate secp256k1 (records signer/manager) and ed25519 (IPNS) keys from same signature request
setup signer and ipns hash in same tx using `fastSetup(..)` function in resolver (rename function :p)
sign records with that deterministic secp256k1 key and publish with IPNs keys, so clients don't have to request wallet to sign individual records.

++still have to lookup if there's side-effect from reusing FIPS hash to priv key in 2 different curve.N values.

**Reply by 0xc0de4c0ffee** [reply]
or 2x signature request is good enough?? no more signature requests to wallets.. it's possible for owner keys to sign as optional & we can generate extra individual IPNS/Manager keys if users require one/another.

let A = `Requesting signature for Deterministic IPNS key\n\nENS Domain: ${ens}\nExtradata: 0x${<keccak256(ens, password?password:"", caip10)>}\nSigned By: ${caip10}`

let B = `Requesting signature for Deterministic Off-Chain ENS Records Manager key\n\nENS Domain: ${ens}\nExtradata: 0x${<keccak256(ens, password?password:"", caip10)>}\nSigned By: ${caip10}`

(A) for IPNS & (B) for record manager/signer using same username/ENS domain and password.

**Reply by sshmatrix** [reply]
I'll add comments on the other thread https://github.com/orgs/namesys-eth/discussions/18

**Reply by sshmatrix** [reply]
> let B = `Requesting signature for Deterministic Off-Chain ENS Records Manager key\n\nENS Domain: ${ens}\nExtradata: 0x${<keccak256(ens, password?password:"", caip10)>}\nSigned By: ${caip10}`

Can we change this to short version: 

`... Deterministic Off-Chain ENS Records Manager key` to `... Off-Chain ENS Records Manager`

**Reply by 0xc0de4c0ffee** [reply]
it's ok, we've to settle somewhere.. 

I'd suggest not using "Domain: .." only as it's keyword used for SIWx ABNF validation to requesting domain.com/App..

**Comment by 0xc0de4c0ffee** [comment]
"Username:" or "ENS Domain:"??

**Reply by sshmatrix** [reply]
Either way is good :P
